ntb_hw_switchtec.c

```diff
@@ -1041,6 +1041,61 @@ static int config_req_id_table(struct switchtec_ntb *sndev,
 	return 0;
 }
 
+static int add_req_ids(struct switchtec_ntb *sndev,
```

Comment:
This seems unnecessarily complicated just to add requester ids... We shouldn't need to allocate a chunk of memory to read back the entire list just to add to that list just to maintain the existing config_req_id_table() interface. And based on the current usage of config_req_id_table() we probably don't a method to add batch ids.

So, roughly, just have three functions:

* add_req_id() - which config_req_id_table() can be change to which just finds the first un-enabled entry on hardware and adds it

* del_req_id() which searches for an entry and disables it

* clr_req_ids() (if necessary) to remove all entries in the table during initialization.

Reply:
The reason I wanted to keep the config_req_id_table() as is is that I wanted to make the lock/unlock of NT operations just in this function, so that other functions like add, del can just focus on the requester ids handling without duplicating those NT operation code. With current code, the config_req_id_table() just program the table while other functions prepare the table. 

For adding a requester id, to prevent from adding duplicated one, it's necessary to go through the entire table and compare each enabled entry with the one to be added. So I think if we can sort the list of enabled entries when we go through them will benefit delete operations a bit, as it will only need to look at the first enabled ones.

And one benefit of addding batch ids with current code is that we can tell if this batch adding will success before we try insert each entry and finally find that there's not enough room for the remaining entries when we have inserted some.

Reply:
I don't think we should worry too much about being efficient about lock and unlock operations as these are all on the slow path. It's certainly not worth the extra mess.

If you do want to combine the lock around the two cases that do a small amount of batch adding, then add a lock and unlock function that must be called around add_req_id and del_req_id().